### DISCUSSION TITLE
Consider resolved queries that take less than X ms as "resolved by cache"

### DISCUSSION THREAD

**Discussion by valantur** [AdguardTeam/AdGuardHome#4744] [Ideas]
I don't use AdGuard's cache and rely on a local Unbound recursive DNS server to do all caching.
However, I love AdGuard's other features and its UI and make decisions based on the information presented there. Therefore, is in my best interest that the info I see on AdGuard's UI is as consistent and accurate as possible.

Right now, if I want to know if a certain query was resolved by my resolver's cache or not, I have to check if the response time was less than 1ms. That tells me my AdGuard instance asked Unbound and Unbound replied from cache.

It would be nice to have an option named "Consider responses that take less than X ms to be returned from cache". This will ensure that Adguard reports accurate information. Even better, you could create a new category of response called something like "Returned from local upstream servers cache" that would let users differentiate between what system's cache is being used, but at the same time consider these responses as cached responses instead of forwarded responses.

Sorry for the wall of text... I hope this makes sense to the Devs.

Thanks!

**Comment by fernvenue** [comment]
Hi @valantur, first, make sure you are talking about AdGuardHome, AdGuardHome and AdGuard are totally different things. Second, AdGuardHome already tells us that these requests are coming from cache or not in response details:

![screenshot](https://user-images.githubusercontent.com/84565547/179070227-240d4a61-9818-40a3-ae09-18e32d24e257.png)

And actually, you can disable cache by setting `cache_size` to `0`, hope this can help you :)

**Reply by cyb3rz3us** [reply]
It looks like you misunderstood the OP's use-case & request - allow me to clarify:

- First, the OP stated that they do not use AG for caching and that `unbound` is used instead.
- Next, the OP states that if a query is serviced in Xms or less, it has been serviced by `unbound`'s cache rather than a recursive lookup.
- Third, the OP is asking for a new feature that will allow the user to set a time threshold such that any response rcv'd <= that time is considered as being served from the upstream's cache.

**Reply by fernvenue** [reply]
Hi @cyb3rz3us, and thanks for your clarification.

> First, the OP stated that they do not use AG for caching and that unbound is used instead.

Yes, so I told him how to disable cache on AdGuardHome.

> Next, the OP states that if a query is serviced in Xms or less, it has been serviced by unbound's cache rather than a recursive lookup.

Actually, in addition to short-term responses not necessarily coming from the cache, very long responses are not necessarily not from the cache. So actually, AdGuardHome can only tell us the response from its own cache or not, and it's already doing that, just like the screenshot I post above.

> Third, the OP is asking for a new feature that will allow the user to set a time threshold such that any response rcv'd <= that time is considered as being served from the upstream's cache.

Like I just said, it is unreasonable to detect whether a response came from a cache by response time, but yeah, we do need some way to filtering query log by response time, and we already have issues related to this: https://github.com/AdguardTeam/AdGuardHome/issues/4050, https://github.com/AdguardTeam/AdGuardHome/issues/2290.

Thanks again for your clarification, you helped me to express more clearly what I meant.

**Reply by valantur** [reply]
thank you @cyb3rz3us

**Reply by valantur** [reply]
@fernvenue you had a really hard time understanding my first post and now you're looking for a hill to die on. Kinda funny.

First of all, I opened this topic in AdGuardHome's repo. Not AdGuardForWindows or any other.

Secondly, AGH's cache is already disabled. I said so in the OP, so I don't know what point you're trying to make and remake with this. I said Unbound was doing all the caching, and the only way to ensure that's the case is by disabling AGH's caching.

Thirdly, what I proposed is having a user-configurable threshold for determining whether a response was actually served from cache (ANY cache, not only AGH's). In my case, I know for a fact that responses that take less than 1ms are served from a local cache. I'd like AGH to represent that instead of saying "it was forwarded to an upstream server and it took 0.11ms).

Also, why is it unreasonable to detect whether a response came from cache by response time? As the Dude would say, that's just your opinion... man.

**Reply by fernvenue** [reply]
> why is it unreasonable to detect whether a response came from cache by response time?

Because:

> Actually, in addition to short-term responses not necessarily coming from the cache, very long responses are not necessarily not from the cache.

So when the upstream server response time is greater than the time you specified, how can you be sure that the response must not come from the cache? Couldn't it be caused by network fluctuations? Couldn't it be that it took more time to establish the connection this time? Couldn't it be that the server is overloaded and the response is slower this time? And you called this option "resolved by cache", how do you ensure that this doesn't mislead users when AdGuardHome having no definitive way to determine its conclusion? I'm just repeating what I've already expressed, and you had a really hard time understanding that, right?

**Reply by cyb3rz3us** [reply]
@fernvenue --- All of the issues you mention are definitely possible but they are not relevant to this discussion - here's why...

You are taking an abstract set of possibilities and applying them as fact to a specific use-case and scenario about which you are unfamiliar.  It isn't appropriate for anyone, except for the network admins, to posit concrete conclusions about their use-case and environment.

The OP is not making a universal absolute pronouncement about all networks.  They are asking for a new feature that will aid in tracking\logging a behavior they have decided is relevant to their specific use-case and scenario.  As long as the implementation allows the feature to be enabled\disabled and to tune the time threshold, I would agree this is probably a useful feature to have.

**Reply by fernvenue** [reply]
> I would agree this is probably a useful feature to have.

@cyb3rz3us But I don't, so I posted my opinion.

AdGuardHome currently gives "served from cache" based on the fact that AdGuardHome does return the response from the cache, not based on some magic correlation that might exist between response time and caching. In this case, telling the user a guess is misleading the user.

At the same time, I do agree with you and @valantur that we need some way to filtering query log by response time.

**Reply by valantur** [reply]
@fernvenue everyone's entitled to their opinion. I'm not criticizing yours, and the option can be easily turned on or off depending of use-case scenario. You don't need that feature, then fine, don't use it. Turn it off. Even better, have the option be off by default.

> So when the upstream server response time is greater than the time you specified, how can you be sure that the response must not come from the cache?. > 

Read below

> Couldn't it be caused by network fluctuations? Couldn't it be that it took more time to establish the connection this time? Couldn't it be that the server is overloaded and the response is slower this time? >

No, I run Unbound on the same machine and monitor its response times and resource usage separately, in an automated and constant manner.

I don't have anything else to add to this. You don't want to see this implemented obviously, and I'm not trying to change your mind. Just out of politeness next time you can try stating your opinion instead of shutting down that of others, based on your own assumptions of "acceptable scenarios", like @cyb3rz3us just said.

**Reply by fernvenue** [reply]
@valantur Yes, but even though it's optional, it still misleads the user into thinking that with the option turned on they will know exactly what the upstream is doing, we know this is impossible, but they don't know, right? All I wanna say is that "resolved by cache" is too much, it must be fact based. 

> ...next time you can try stating your opinion instead of shutting down that of others...

I respect your opinion, I also think it's essential to point out my concerns. And thanks for @cyb3rz3us's clarification I understand more of your point, and I've made my own expression more precise. 

> ...we do need some way to filtering query log by response time, and we already have issues related to this: https://github.com/AdguardTeam/AdGuardHome/issues/4050, https://github.com/AdguardTeam/AdGuardHome/issues/2290.

What do you guys think about this way? I think this is actually a better solution.

**Reply by cyb3rz3us** [reply]
> we know this is impossible, but they don't know, right

As I mentioned elsewhere, as long as the feature is well documented, the admin should know this is a presumption and not 100% deterministic.  If they don't, then they really shouldn't be that network's admin.

> filtering query log by response time...actually a better solution

No - it's a solution to a different issue.  Whether or not it satisfies @valantur's need is for them to say.  Do they want to get a separate log message?  Do they want a counter?  Both or something else entirely.

Personally, I would want the log entry to be notated with something like the following:

"Served by external cache (user cfg)"

**Reply by fernvenue** [reply]
> As I mentioned elsewhere, as long as the feature is well documented, the admin should know this is a presumption and not 100% deterministic. If they don't, then they really shouldn't be that network's admin.

Imagine if a calculator gave an exact answer like "resolved by cache", but wrote in the corner of the manual that it might give a very wrong answer, would you still think that the user should not be a user? I really can't understand how a judgment based on a magical possibility can be a boolean.

**Reply by cyb3rz3us** [reply]
Actually calculators do give "wrong" answers all the time such as whenever one uses Pi or the natural log.  Both are non-repeating numbers and consequently, the calculator uses an approximation.  The same is also true with any number of trig functions.

My point is this: presumptions, approximations, etc. are perfectly reasonable provided all of the interested parties are aware that the presumptions, approximations, etc. are being used and the limitations of same.

Thank you for making my point...

**Reply by fernvenue** [reply]
> the calculator provides an approximation

Do you really think the exact opposite conclusion is an approximation? In the case we are discussing, AdGuardHome may give "resolved by cache" when the upstream server using recursive lookup, vice versa.

**Reply by cyb3rz3us** [reply]
Seriously man - this has gone on long enough.  The endless contrarian view with zero substantive rationale serves no useful purpose. The case for the feature has been made - and made quite well - and it's now up to the devs...

**Reply by fernvenue** [reply]
> First of all, I opened this topic in AdGuardHome's repo. Not AdGuardForWindows or any other.

I forgot this, some people really don't know the difference between them, like https://github.com/AdguardTeam/AdGuardHome/issues/4712 this one. So I'm just make sure you are using AdGuardHome.

**Reply by fernvenue** [reply]
> The endless contrarian view with zero substantive rationale serves no useful purpose.

I've given enough probabilities that the truth could be quite the opposite, let me quote it again for you:

> So when the upstream server response time is greater than the time you specified, how can you be sure that the response must not come from the cache? Couldn't it be caused by network fluctuations? Couldn't it be that it took more time to establish the connection this time? Couldn't it be that the server is overloaded and the response is slower this time? And you called this option "resolved by cache", how do you ensure that this doesn't mislead users when AdGuardHome having no definitive way to determine its conclusion? 

And if you are talking about yourself, then yes, I really don't see you giving any substantial reason why it's reasonable to tell users the exact opposite conclusion.

**Comment by cyb3rz3us** [comment]
> telling the user a guess is misleading the user.

Actually, as long as the feature is well defined and documented, it isn't misleading at all.  If a user decides to not RTFM & misunderstands the feature, that's not being misled - that's willful ignorance.  Can you imagine if we opted to not build features because some users didn't RTFM?  We'd never have any features...

As an aside, it is puzzling (to me at least) that @fernvenue is in favor of filtering query responses based on response time but is against having a feature akin to the following:

---

#### External cache tracking

- [ ] Consider & log all query responses equal to or below X milliseconds as served from external cache.

**Reply by fernvenue** [reply]
> Actually, as long as the feature is well defined and documented, it isn't misleading at all.

I agree with you, but "resolved by cache" is clearly not well-defined, "resolved by cache" is too much, it must be fact based, not a possibility.

**Reply by cyb3rz3us** [reply]
> it must be fact based

We can agree to disagree here but regardless, that would be a question for the devs.  Are they OK with having a feature that presents data based on a clearly understood presumption or only fact based.  Both methods are reasonable depending on the situation.

**Reply by fernvenue** [reply]
> Both methods are reasonable depending on the situation.

So tell me yes or no: "resolved by cache" is well-defined or not? If your answer is no, why "...Both methods are reasonable..." ?

**Comment by ainar-g** [comment]
I don't really feel like this is a feature that should be added to AdGuard Home.  Low query durations do not necessarily mean that they were resolved from the upstream cache.  And conversely, query durations above the threshold don't mean that the queries *weren't* resolved from cache.  I feel like this kind of statistic is better off in a custom query log database, which will be possible once #3389 is resolved.

**Reply by fernvenue** [reply]
Totally agree.
